src/provider.toit

```diff
@@ -0,0 +1,166 @@
+// Copyright (C) 2023 Toitware ApS. All rights reserved.
+// Use of this source code is governed by an MIT-style license that can be found
+// in the LICENSE file.
+
+import esp32
+import log
+import monitor
+import system.services show ServiceProvider ServiceResource ServiceHandler
+import .service
+
+class WatchdogServiceProvider extends ServiceProvider
+  implements ServiceHandler:
+
+  static GRANULARITY-MS ::= 2_000
+
+  dogs_/Map ::= {:}  // From string to Watchdog.
+  system-watchdog-task_/Task? := null
+  logger_/log.Logger := (log.default.with-name "watchdog").with-level log.INFO-LEVEL
+  mutex_/monitor.Mutex := monitor.Mutex
+
+  constructor:
+    super "watchdog" --major=1 --minor=0
+    provides WatchdogService.SELECTOR --handler=this
+
+  handle index/int arguments/any --gid/int --client/int -> any:
+    if index == WatchdogService.CREATE-INDEX:
+      id := arguments[0] as string
+      dog/Watchdog := ?
+      if dogs_.contains id:
+        // There exists already a watchdog with this id.
+        // Remove it and replace it with a fresh one.
+        // However, the new watchdog will be in the same state as the old one.
+        old-dog := dogs_[id]
+        dogs_.remove id
+        dog = Watchdog.clone this client old-dog
+      else:
+        dog = Watchdog this client id
+      dogs_[id] = dog
+      return dog
+
+    if index == WatchdogService.START-INDEX:
+      dog := (this.resource client arguments[0]) as Watchdog
+      dog.start (arguments[1] as int)
+      start-system-watchdog-if-necessary_
+      return null
+
+    if index == WatchdogService.FEED-INDEX:
+      dog := (this.resource client arguments[0]) as Watchdog
+      dog.feed
+      return null
+
+    if index == WatchdogService.STOP-INDEX:
+      dog := (this.resource client arguments[0]) as Watchdog
+      dog.stop
+      stop-system-watchdog-if-possible_
+      return null
+
+    unreachable
+
+  start-system-watchdog-if-necessary_:
+    if system-watchdog-task_: return
+
+    mutex_.do:
+      logger_.debug "starting system watchdog"
+      esp32.watchdog-init --ms=GRANULARITY-MS
+      system-watchdog-task_ = task::
+        while true:
+          too-late := false
+          dogs_.do --values: | dog/Watchdog |
+            if dog.is-too-late:
+              logger_.fatal "watchdog too late" --tags={ "id": dog.id }
+              too-late = true
+
+          if too-late:
+            esp32.deep-sleep Duration.ZERO
+          else:
+            // Feed the system watchdog.
+            logger_.debug "feeding system watchdog"
+            mutex_.do: esp32.watchdog-reset
+            sleep --ms=1_000
```

Comment:
Shouldn't this be based on GRANULARITY-MS?

Reply:
done.